agere2015-akka-js/agere-akkajs.tex

```diff
@@ -147,20 +150,40 @@
 
 \section{Introduction}
 
-In recent times, there has been a significant pique of interest regarding JavaScript. The language, originally developed by Brendan Eich, while working at Netscape, and once confined to browsers, is now being used to power a wide variety of projects, ranging from server side applications \cite{web:nodejs} to hardware \cite{web:tessel} to robots \cite{web:johnnyfive}.
+Since it introduction in 1995~\cite{} there has been an increasing interest on JavaScript. Originally, this programming language was developed by Netscape Inc. to provide a lightweight interpreted language to captivate non-professional programmers. Later, thanks to a standardisation process started in 1996 with the first release~\cite{} and, that continues nowadays with the sixth version released in June 2015~\cite{}, Javascript has become one of the most popular programming languages on the Web.
 
+It is interesting to analyze the evolution of the language. Before the introduction of Ajax~\cite{} Javascript was principally used to do scripting in the Browser. In 2009 CommonJS~\cite{} and Node.js~\cite{web:nodejs} proposed an ecosystem for running Javascript outside the browser. With the rise of single-page web applications~\cite{} and source-to-source compilers~\cite{}, Javascript is now being used to power a large variety of projects. 
 
-With the rise of this programming language, a need has emerged for a pattern which can easily manage concurrent and distributed scenarios. Current approaches in the JavaScript world include the usage of callbacks and promises: the former of which has the unfortunate side-effect of forcing inversion of control, while the latter has no built-in solution for distributed computing.
+Despite the advances made by the javascript community, there are several issues that make developing software using Javascript harder. Working in Javascript is inconvenient because of its module system (only in ES6~\cite{}), its weak-typing, the verbose function syntax\footnote{Its support for closures is commonly noted as being one of JavaScript\rq{s} redeeming features.}, the late binding\footnote{Early binding allows for static verification of the existence of method-signature pairs}, the lack of static types, the \emph{this} behaviour, the equality/automatic conversion to boolean and the asynchronous programming heavily based on callbacks.

```

Comment:
I agree it’s good to point out the issues of async programming with JavaScript (maybe slightly deemphasize issues not related to async programming).

However, I would try to avoid saying that all existing approaches (for async prog) are “bad”, since it’s almost impossible to “prove” such a statement (or at least substantiate it).

I think we can simplify our lives using the following line of argument:

1. Async programming in JavaScript is difficult (all I/O and UI events are async, callback hell etc.)
2. The actor model has been used successfully to build highly-concurrent and distributed systems across several domains (telecom [Ericsson], web commerce [Amazon S3], etc.).
3. Unfortunately, many widely-used actor programming systems and libraries do not support deploying actor-based programs on JavaScript.
4. For example, Akka is a widely-used actor framework with APIs for both Java and Scala.
5. However, as it stands, targeting JavaScript platforms is impossible, despite tools like Scala.js, the Scala-to-JavaScript compiler.
6. The main reason is that the Akka runtime system relies on a variety of JVM-specific features and libraries.
7. This papers presents Akka.js... it allows compiling actor-based programs in Scala to run on JavaScript using the same API as Akka-based programs targeting the JVM...